src/common/error_code.cc

```diff
@@ -0,0 +1,188 @@
+using boost::system::error_condition;
+using boost::system::generic_category;
+using boost::system::system_category;
+
+namespace ceph {
+
+// A category for error conditions particular to Ceph
+
+class ceph_error_category : public converting_category {
+public:
+  ceph_error_category(){}
+  const char* name() const noexcept override;
+  std::string message(int ev) const override;
+  using converting_category::equivalent;
+  bool equivalent(const boost::system::error_code& c,
+		  int ev) const noexcept override;
+  int from_code(int ev) const noexcept override;
+};
+
+const char* ceph_error_category::name() const noexcept {
+  return "ceph";
+}
+
+std::string ceph_error_category::message(int ev) const {
+  if (ev == 0)
+    return "No error";
+
+  switch (static_cast<errc>(ev)) {
+
+  case errc::not_in_map:
+    return "Map does not contain requested entry.";
+  case errc::does_not_exist:
+    return "Item does not exist";
+  case errc::failure:
+    return "An internal fault or inconsistency occurred";
+  case errc::exists:
+    return "Already exists";
+  case errc::limit_exceeded:
+    return "Attempt to use too much";
+  case errc::auth:
+    return "Authentication error";
+  case errc::conflict:
+    return "Conflict detected or precondition failed";
+  }
+
+  return "Unknown error.";
+}
+
+bool ceph_error_category::equivalent(const boost::system::error_code& c,
+				     int ev) const noexcept {
+  if (c.category() == system_category()) {
+    if (c.value() == boost::system::errc::no_such_file_or_directory) {
+      if (ev == static_cast<int>(errc::not_in_map) ||
+	  ev == static_cast<int>(errc::does_not_exist)) {
+	// Blargh. A bunch of stuff returns ENOENT now, so just to be safe.
+	return true;
+      }
+    }
+    if (c.value() == boost::system::errc::io_error) {
+      if (ev == static_cast<int>(errc::failure)) {
+	return true;
+      }
+    }
+    if (c.value() == boost::system::errc::file_exists) {
+      if (ev == static_cast<int>(errc::exists)) {
+	return true;
+      }
+    }
+    if (c.value() == boost::system::errc::no_space_on_device ||
+	c.value() == boost::system::errc::invalid_argument) {
+      if (ev == static_cast<int>(errc::limit_exceeded)) {
+	return true;
+      }
+    }
+    if (c.value() == boost::system::errc::operation_not_permitted ||
+	c.value() == boost::system::errc::permission_denied) {
+      if (ev == static_cast<int>(errc::limit_exceeded)) {
+	return true;
+      }
+    }
+    if (c.value() == boost::system::errc::operation_not_permitted) {
+      if (ev == static_cast<int>(ceph::errc::conflict)) {
+	return true;
+      }
+    }
+  }
+  return false;
+}
+
+int ceph_error_category::from_code(int ev) const noexcept {
+  if (ev == 0)
+    return 0;
+
+  switch (static_cast<errc>(ev)) {
+  case errc::not_in_map:
+  case errc::does_not_exist:
+    // What we use now.
+    return -ENOENT;
+  case errc::failure:
+    return -EIO;
+  case errc::exists:
+    return -EEXIST;
+  case errc::limit_exceeded:
```

Comment:
this doesn't match equivalent() above -- limit_exceeded and conflict are different there

Reply:
Fixed.